### PR TITLE
Avoid intersection by transit through Start

### PR DESCRIPTION
**Pull request by balghane**
This PR adds a concept to the Lift of a current and desired Position. It also considers whether positions are above or below the starting position. When moving from above to below or vice versa, the robot will transit through the starting position, which avoids any possibility of self-intersection.

### CODE REVIEW

**Comment by balghane** [comment]
Whoops also need to allow going straight to desired when we don't have a game piece

File: src/main/java/frc/robot/Cal.java [hunk 1]
```diff
@@ -109,6 +109,17 @@ public static final class Lift {
...
+    /**
+     * Thresholds for when we consider the lift has reached a position. This is logical (for
+     * considering where the lift can go) but not functional (does not stop arm control or something
+     * when reached). We apply broader thresholds for the Starting position as the lift transits
+     * through this position.
+     */
+    public static final double ELEVATOR_THRESHOLD_INCHES = 0.5,
+        ARM_THRESHOLD_DEGREES = 2.0,
+        ELEVATOR_START_THRESHOLD_INCHES = 0.5,
+        ARM_START_THRESHOLD_DEGREES = 2.0;
```
Comment (tails618) [inline src/main/java/frc/robot/Cal.java]: Either starting position thresholds should be larger, like the comment says, or that line in the comment should be removed. If we don't know how much larger they should be, make it `0.5 + PLACEHOLDER_DOUBLE`.

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 2]
```diff
@@ -246,9 +255,76 @@ public void initialize() {
...
+    return true;
+  }
+
+  /** Converts a lift position to a relative lift position (above or below starting position). */
+  private static LiftPositionStartRelative getRelativeLiftPosition(LiftPosition pos) {
+    switch (pos) {
+      case STARTING:
+        return LiftPositionStartRelative.AT_START;
+      case SCORE_MID:
+      case SCORE_HIGH:
```
Comment (tails618) [inline src/main/java/frc/robot/subsystems/Lift.java:295]: Why are these empty?

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 3]
```diff
@@ -246,9 +255,76 @@ public void initialize() {
...
+    if (atPosition(LiftPosition.STARTING)) {
+      latestPosition = LiftPosition.STARTING;
+    } else if (atPosition(desiredPosition)) {
+      latestPosition = desiredPosition;
+    }
+
+    // If the lift is going from below to above or above to below, we have to transit through
+    // starting position. Otherwise, we can go directly to the desired position.
+    LiftPositionStartRelative latestPositionStartRelative = getRelativeLiftPosition(latestPosition);
+    if (getRelativeLiftPosition(desiredPosition) == LiftPositionStartRelative.BELOW_START
```
Comment (tails618) [inline src/main/java/frc/robot/subsystems/Lift.java]: Why do we run `getRelativeLiftPosition(desiredPosition)` in the if statement, but `getRelativeLiftPosition(latestPosition)` is its own variable? I would change the relative desired position into its own variable as well; it's more readable, and it is used multiple times.

**Review by tails618 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 2 above)
Comment (balghane) [reply src/main/java/frc/robot/subsystems/Lift.java:295]: It's a "fallthrough." If you don't break, the switch continues on to the code in the next case. Sometimes it's considered an anti-pattern but here I think it's okay.

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 3 above)
Comment (balghane) [reply src/main/java/frc/robot/subsystems/Lift.java]: Changed

File: src/main/java/frc/robot/Cal.java (hunk 1 above)
Comment (balghane) [reply src/main/java/frc/robot/Cal.java]: Sure, I doubled them, we will almost certainly want to do more tuning later but this is a reasonable start

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 2 above)
Comment (tails618) [reply src/main/java/frc/robot/subsystems/Lift.java:295]: Ok. I just read an article about it and I think I understand it now.

**Review by tails618 — APPROVED** [review]

**Pull request merged**